Improve DataCatalog 2.0 docstring

ravi-kumar-pilla opened the pull request:
## Description
Resolves https://github.com/kedro-org/kedro/issues/4744

## Development notes
- Updated docstrings for `DataCatalog`, `SharedMemoryDataCatalog`, `CatalogConfigResolver` and `CatalogCommandsMixin`


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [x] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

ravi-kumar-pilla commented on kedro/framework/context/catalog_mixins.py:
```diff
@@ -9,14 +9,86 @@
+
+    Methods:
+        - list_datasets: Show datasets per type for specified pipelines.
+        - list_patterns: List all dataset factories in the catalog.
```
Yes I agree. It started as factory patterns. Lets call them `patterns` everywhere

ravi-kumar-pilla commented on kedro/io/catalog_config_resolver.py:
```diff
@@ -209,27 +368,125 @@ def list_patterns(self) -> list[str]:
+        Match a dataset name against the user-defined catch-all pattern.
+
+        This method checks if the given dataset name matches any of the user-defined
+        catch-all patterns. If a match is found, the first matching pattern is returned.
```
okay, the `next` method always returns the first item in the iterator right ? Also I don't understand how can a user provide multiple catch-all patterns, since we have a validation which leads to -
```bash
DatasetError: Multiple catch-all patterns found in the catalog: {name2}, {name}. Only one catch-all pattern 
is allowed, remove the extras.
```

ElenaKhaustova replied:
Yes, you're right - my bet

ravi-kumar-pilla commented on kedro/io/catalog_config_resolver.py:
```diff
@@ -243,7 +500,35 @@ def _extract_patterns(
+
+        This method identifies dataset patterns from the catalog configuration and
+        sorts them based on specificity. It also identifies a catch-all pattern (if any)
+        and sets it as the default pattern.
```
I forgot to rename some parts after merging the latest PR with updated names. Sorry for that

ravi-kumar-pilla replied:
I think the statement here makes sense. Like if catch-all pattern is available then that would be the default pattern

merelcht reviewed (commented):
This looks great and very thorough!! 

I'm just curious why the mention of dataset factory was removed? It's still the name of the feature: https://docs.kedro.org/en/stable/data/kedro_dataset_factories.html, so might actually help to keep the mentions in the code.

ravi-kumar-pilla commented:
> I'm just curious why the mention of dataset factory was removed? It's still the name of the feature: https://docs.kedro.org/en/stable/data/kedro_dataset_factories.html, so might actually help to keep the mentions in the code.

Thank you for reviewing @merelcht ... It felt inconsistent where we name factory some times and patterns some times - https://github.com/kedro-org/kedro/pull/4751#discussion_r2098231362

I thought may be we can be consistent by avoiding factory naming instead use `dataset patterns` everywhere instead of `factory patterns` or `dataset factory patterns`. If this will confuse users more, happy to revert some changes.

merelcht commented:
> I thought may be we can be consistent by avoiding factory naming instead use `dataset patterns` everywhere instead of `factory patterns` or `dataset factory patterns`. If this will confuse users more, happy to revert some changes.

It's the name we've used for the feature since the beginning, so I think it will be very confusing if we suddenly don't use that anymore when talking about this same feature. 

Naming decisions like this are quite significant, and if we're considering a change, I think it’s important to have a proper discussion with the team rather than including it in a larger PR where it might get overlooked.

479BetaCoder commented:
> Naming decisions like this are quite significant, and if we're considering a change, I think it’s important to have a proper discussion with the team rather than including it in a larger PR where it might get overlooked.

I agree. Reverted those changes and added `dataset factory patterns` in the docstrings where needed. The variable name we have in the class is `_dataset_patterns`.  I did not change that.

ravi-kumar-pilla commented:
> Naming decisions like this are quite significant, and if we're considering a change, I think it’s important to have a proper discussion with the team rather than including it in a larger PR where it might get overlooked.

I agree. Reverted those changes and added dataset factory patterns in the docstrings where needed. The variable name we have in the class is _dataset_patterns. I did not change that. Thank you
